python/mxnet/contrib/amp/amp.py

```diff
@@ -342,3 +349,320 @@ def unscale(optimizer_or_trainer):
+        FP32 HybridBlock or SymbolBlock object
+    target_dtype : str or numpy
+        currently only supports fp16. The target dtype indicates to add cast layers
+        when possible so that lower precision computation can be leveraged.
+    target_precision_ops : list of strs
+        Override the list of operator names casted to target_dtype.
+        If None, uses the framework's default list to be casted to FP32.
+    conditional_fp32_ops : list of (str, str, list of str)
+        Override the list of functions to be casted to FP32.
+        The format of the list is
+        (name of the function, name of the parameter,
+         list of values of the parameter that make the operator to be casted to FP32
+    excluded_sym_names : list of strs
+        A list of strings that represent the names of symbols that users want to exclude
+        from being quantized
+    ctx : Context
+        Context on which model parameters should live
+    """
+    from ...gluon import HybridBlock, SymbolBlock
+    assert isinstance(block, HybridBlock), "block input should be a HybridBlock"
+    if not block._cached_graph:
+        raise RuntimeError(
+            "Please first call block.hybridize() and then run forward with "
+            "this block at least once before calling export.")
+
+    # Prepare inputs to pass to the convert_symbol API
+    inputs, sym = block._cached_graph
+    input_names = []
+    for inp in inputs:
+        input_names.append(inp.name)
+    converted_sym = convert_symbol(sym, target_dtype, target_dtype_ops,
+                                   fp32_ops, conditional_fp32_ops,
+                                   excluded_sym_names, data_names=input_names)
+
+    arg_names = set(converted_sym.list_arguments())
+    aux_names = set(converted_sym.list_auxiliary_states())
+    arg_dict = {}
+
+    # If dtype for the param was set in the json, cast the
+    # param to this dtype
+    attr_dict = converted_sym.attr_dict()
+    for name, param in block.collect_params().items():
+        if name in arg_names:
+            arg_dict['arg:%s'%name] = param._reduce()
+            if name in attr_dict and "__dtype__" in attr_dict[name]:
+                if attr_dict[name]["__dtype__"] != "-1":
+                    typ = _DTYPE_MX_TO_NP[int(attr_dict[name]["__dtype__"])]
+                    arg_dict['arg:%s'%name] = arg_dict['arg:%s'%name].astype(typ)
+        else:
+            assert name in aux_names
+            arg_dict['aux:%s'%name] = param._reduce()
+            if name in attr_dict and "__dtype__" in attr_dict[name]:
+                if attr_dict[name]["__dtype__"] != "-1":
+                    typ = _DTYPE_MX_TO_NP[int(attr_dict[name]["__dtype__"])]
+                    arg_dict['aux:%s'%name] = arg_dict['aux:%s'%name].astype(typ)
+
+    # Create a symbolblock and cast the params to the dtypes based
+    # on the dtype information from the converted_symbol
+    ret = SymbolBlock(converted_sym, inputs)
```

Comment:
just curious, why are we using SymbolBlock here?

Reply:
Given a symbol and params, SymbolBlock is supposed to be used to load the model and run inference on it. Other Hybrid blocks are more for the hybrid mode, where you run imperative mode on the first run and cache ops and run forward on cached symbol from the second pass.